enhancements/78-dynamically-reclaiming-resources/README.md

```diff
@@ -0,0 +1,142 @@
+
+### Pod Failure Cases
+
+#### RestartPolicy of Pods as Never
+
+Whenever a Pod of a Job fails, the Job will recreate a Pod as replacement against the failed Pod. The Job will continue this process of creating new Pods until the Job reaches its `backoffLimit`(by default the backoffLimit is `6`). The newly created Pod against the failed Pod will reuse the resources of the failed Pod. Once the Job reaches its `backoffLimit` and does not have the required successful `.spec.completions` count, the Job is termed as a `Failed` Job. When the Job is marked as failed, no more Pods would be created by the Job. So, the remaining owned resources of the Job will be reclaimed by queue. 
+
+#### RestartPolicy of Pods as OnFailure
+
+When the Pods of the Job have `.spec.template.spec.restartPolicy = "OnFailure"`, the failed Pod stays on the node and the failed container within the Pod is restarted. The Pod might run indefinitely or get marked as `Failed` after the `.spec.activeDeadlineSeconds` is completed by Job if specified. In this case also, Kueue won't reclaim the resources of the Job until the Job gets completed as either `Failed` or `Succeeded`.
+
+Hence, as seen from the above discussed Pod failure cases, we conclude that the Pods of a Job which fail during its execution, its resources are not immediately reclaimed by Kueue. Only when the Job gets marks as `Failed`, the resources of failed Pods will be reclaimed by Kueue.
+
+## API
+
+A new field `ReclaimedPodSetPods` is added to the `.status` of Workload API.
+
+```go
+// WorkloadStatus defines the observed state of Workload
+type WorkloadStatus struct {
+    // conditions hold the latest available observations of the Workload
+    // current state.
+    // +optional
+    // +listType=map
+    // +listMapKey=type
+    Conditions []WorkloadCondition `json:"conditions,omitempty"`
+
+    // ReclaimedPodSetPods are the pods (by PodSet name) which have
+    //completed their execution successfully and resources of the pods
+	//are reclaimed.
+    // +optional
+    ReclaimedPodSetPods ReclaimedPodSetPods `json:"reclaimedPodSetPods"`
+}
+
+type ReclaimedPodSetPods map[string]struct{
+    Count int
+    Pods []string
+}
+```
+
+#### ReclaimedPodSetPods
+`.status.reclaimedPodSetPods` holds the mapping between PodSet name to the count of Pods and Pod names that have successfully completed its execution and their resources have been reclaimed by Kueue. The names of Pods are recorded to identify the Pods whose resources have been reclaimed by Kueue. If a name of a Pod already exists in the list then the resources associated with the succeeded Pod have already been reclaimed by Kueue.
+
+## Implementation
+
+The Workload reconciler will keep a watch(or `Watches()`) on the Pods of the Job. The Workload reconciler will process the events of the Pods and update `.status.reclaimedPodSetPods` field accordingly.
```

Comment:
In the case of Job, why not just watch the Job status and observe the `.succeeded` field?

Reply:
Yes, changed the approach and watching the `.succeeded`field of the job status.

Reply:
If we want to support other Jobs, we should watch them all, right? Why not achieve this through job reconciling. Today when job completed, we'll update workloads in Job reconciler, and workload controller will get update events. I think it's similar.

Actually, I prefer the original design, by watching the Pods of Job. This is self closed-loop with kueue. And if other Jobs don't support `.status.succeeded` field, they have to implement it when integrating with kueue.  

But the disadvantage is we don't know whether this is as expected. For batch Job, we know that the succeeded pods can be reclaimed, but for other Jobs, we don't know.

Reply:
We don't even know that for Job, because parallelism can be different from completions. So understanding the specific API is necessary.